### ISSUE TITLE
Kani symbols not available in path dependency when build script depends on same local crate

### ISSUE THREAD

**Issue by xarantolus** [model-checking/kani#3101]
Hey, I am currently running into a compilation error when I have two local crates where one depends on the other in the build script.

TL;DR: `kani` symbols are not available in path dependency crate if dependent crate has a build script that also imports the same path dependency.

This is the directory structure:

```
binary
binary/build.rs
...
constants
```

And binary has a both a build and normal dependency on `constants` (`binary/Cargo.toml`):

```
[dependencies]
constants = { path = "../constants" }

[build-dependencies]
constants = { path = "../constants" }
```

Both crates have some `#[cfg(kani)]` directives in their code. When I run `cargo kani` in `constants`, it compiles and verifies. However, when running `cargo kani` in the `binary` crate, I get this error:

```
$ cd binary && cargo kani -v
Kani Rust Verifier 0.48.0 (cargo plugin)
Skipped the following unsupported targets: 'build-script-build'.
[Kani] Running:

[...]

error[E0433]: failed to resolve: use of undeclared crate or module `kani`
 --> constants/src/lib.rs:9:6
  |
9 | impl kani::Arbitrary for SomeStruct {
  |      ^^^^ use of undeclared crate or module `kani`

error[E0433]: failed to resolve: use of undeclared crate or module `kani`
  --> constants/src/lib.rs:12:25
   |
12 |             some_field: kani::any()
   |                         ^^^^ use of undeclared crate or module `kani`
```

This only happens if the `build.rs` script that also depends on the `constants` crate is present; if you delete it (no need to adjust `Cargo.toml`, just remove the script), everything works again. However, I need the build script for my normal build.

I tried this code: https://github.com/xarantolus/kani-issue-build-script, it contains the mentioned code and structure.

using the following command line invocation:

```
cd binary && cargo kani -v
```

with Kani version: `Kani Rust Verifier 0.48.0 (cargo plugin)`

I expected to see this happen: compilation should work for the dependency crate (`constants`) even when a build script that depends on `constants` is present

Instead, this happened: It seems like kani is not injected into the path dependency `constants` if the package we want to verify has a build script that also depends on the `constants` package.

**Comment by adpaco-aws** [comment]
Hi @xarantolus , thanks for opening this issue! I've been able to reproduce it using your code and steps.

**Comment by tautschnig** [comment]
We can get a bit further by doing the following (for a suitable value of `PATH-TO-KANI`):
````patch
--- a/kani-driver/src/call_cargo.rs
+++ b/kani-driver/src/call_cargo.rs
@@ -326,10 +326,28 @@ pub fn cargo_config_args() -> Vec<OsString> {
     [
         "--target",
         env!("TARGET"),
-        // Propagate `--cfg=kani` to build scripts.
+        // Propagate kani library config to build scripts.
         "-Zhost-config",
         "-Ztarget-applies-to-host",
-        "--config=host.rustflags=[\"--cfg=kani\"]",
+        "--config=host.rustflags=[
+            \"-C\",
+            \"panic=abort\",
+            \"--cfg=kani\",
+            \"-Z\",
+            \"unstable-options\",
+            \"-Z\",
+            \"crate-attr=feature(register_tool)\",
+            \"-Z\",
+            \"crate-attr=register_tool(kanitool)\",
+            \"--sysroot\",
+            \"<PATH-TO-KANI>/target/kani\",
+            \"-L\",
+            \"<PATH-TO-KANI>/target/kani/lib\",
+            \"--extern\",
+            \"kani\",
+            \"--extern\",
+            \"noprelude:std=<PATH-TO-KANI>/target/kani/lib/libstd.rlib\"
+        ]",
     ]
     .map(OsString::from)
     .to_vec()
````
with the following additional observations:
1. This does make the search for the `kani` crate work, but eventually fails linking object files.
2. Some of these options ought to be already passed via `CARGO_ENCODED_RUSTFLAGS`, but this doesn't seem to be picked up.
3. As further indication that `CARGO_ENCODED_RUSTFLAGS` is simply ignored: `--extern kani` requires `-Z unstable-options`, but we don't add the latter to `CARGO_ENCODED_RUSTFLAGS`.
